src/client/renderer/r_atlas.c

```diff
@@ -0,0 +1,237 @@
+		R_BindTexture(atlas->image.texnum);
+	
+	// FIXME: quite possibly the worst stitch algorithm known to man.
+	// stitch and set up some initial data
+	for (uint16_t i = 0; i < atlas->images->len; i++) {
+		r_atlas_image_t *image = &g_array_index(atlas->images, r_atlas_image_t, i);
+
+		image->position[0] = width;
+		image->position[1] = 0;
+
+		width += image->input_image->width;
+		height = MAX(height, image->input_image->height);
+
+		min_size = MIN(min_size, MIN(image->input_image->width, image->input_image->height));
+		
+		image->image.width = image->input_image->width;
+		image->image.height = image->input_image->height;
+		image->image.type = IT_ATLAS_IMAGE;
+		image->image.texnum = atlas->image.texnum;
+
+		g_hash_table_replace(atlas->hash, (gpointer) image->input_image, (gpointer) image);
+	}
+
+	// see how many mips we need to make
+	uint16_t num_mips = 1;
+
+	while (min_size > 1) {
+		min_size = floor(min_size / 2.0);
+		num_mips++;
+	}
+
+	// set up num mipmaps
+	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_BASE_LEVEL, 0);
+	glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAX_LEVEL, num_mips - 1);
+
+	// make all of the mips as we go
+	for (uint16_t i = 0; i < num_mips; i++) {
+		const r_pixel_t mip_width = width >> i;
+		const r_pixel_t mip_height = height >> i;
+
+		SDL_Surface *mip_surface = SDL_CreateRGBSurface(0, mip_width, mip_height, 32, RMASK, GMASK, BMASK, AMASK);
+
+		if (!mip_surface)
+			Com_Error(ERR_DROP, "Unable to allocate memory for atlas");
+
+		// pop in all of the textures
+		for (uint16_t j = 0, w = 0; j < atlas->images->len; j++) {
+			const r_atlas_image_t *image = &g_array_index(atlas->images, r_atlas_image_t, j);
+			SDL_Surface *image_surface;
+			GLint image_mip_width;
+			GLint image_mip_height;
+
+			R_BindTexture(image->input_image->texnum);
+			
+			// this has to be done, otherwise mipmaps for the image specified may not have been generated yet.
+			// these actually force it to generate the mip.
+			glGetTexLevelParameteriv(GL_TEXTURE_2D, i, GL_TEXTURE_WIDTH, &image_mip_width);
+			glGetTexLevelParameteriv(GL_TEXTURE_2D, i, GL_TEXTURE_HEIGHT, &image_mip_height);
+
+			image_surface = SDL_CreateRGBSurface(0, image_mip_width, image_mip_height, 32, RMASK, GMASK, BMASK, AMASK);
+
+			SDL_SetSurfaceBlendMode(image_surface, SDL_BLENDMODE_NONE);
+
+			SDL_LockSurface(image_surface);
+
+			glGetTexImage(GL_TEXTURE_2D, i, GL_RGBA, GL_UNSIGNED_BYTE, image_surface->pixels);
+
+			SDL_UnlockSurface(image_surface);
+
+			SDL_Rect dst = {
+				.x = w,
+				.y = 0,
+				.w = image_mip_width,
+				.h = image_mip_height
+			};
+
+			SDL_BlitSurface(image_surface, NULL, mip_surface, &dst);
+
+			SDL_FreeSurface(image_surface);
+
+			w += image_mip_width;
+		}
+
+		SDL_LockSurface(mip_surface);
+
+		R_BindTexture(atlas->image.texnum);
+
+		glTexImage2D(GL_TEXTURE_2D, i, GL_RGBA, mip_width, mip_height, 0, GL_RGBA, GL_UNSIGNED_BYTE, mip_surface->pixels);
+
+		SDL_UnlockSurface(mip_surface);
+
+		SDL_FreeSurface(mip_surface);
```

Comment:
Damn. I'm slightly surprised that this much work is necessary here. Just curious: why can't we simple rely on glGenerateMipmap?